### ISSUE TITLE
Powerline or colors wrong?

### ISSUE THREAD

**Issue by ddulic** [gpakosz/.tmux#40]
![screenshot from 2016-11-23 09-16-42](https://cloud.githubusercontent.com/assets/10076428/20554667/bdb7f1a6-b15d-11e6-93c1-2b73ac6ade60.png)

What am I doing wrong :O ? Installed powerline and it looks like this :/ And why don't I have any colors in my bash?
Thanks for the conf file!

**Comment by gpakosz** [comment]
Hello,

Quit tmux and then make sure make sure `TERM` is set to `xterm-256color`.

**Comment by ddulic** [comment]
Its set to the correct value. I am using Gnome Drop Down Terminal, but the problem also occurs in Terminator.
Thanks for the fast response!

**Comment by ddulic** [comment]
I tried removing tmux from the option "Run a custom command instead of my shell" in both terminals and when I manually run tmux the display and colors are as intended. Is this normal behavior? Any known workaround?

![screenshot from 2016-11-23 10-41-40](https://cloud.githubusercontent.com/assets/10076428/20556910/7985a724-b169-11e6-85d1-c25ae0117e08.png)

**Comment by gpakosz** [comment]
I'm not using Gnome Drop Down Terminal nor Terminator.

Though if you just relaunched tmux and you're getting a session numbered `9` then definitely all tmux processes were not killed. Try to kill all and start over?

In any case to me it's all about setting `TERM` correctly outside of tmux.

**Comment by ddulic** [comment]
After a bit of messing around the problem seems to be that the value of $TERM isn't being correctly passed to the terminal when running the command instead of the shell. For terminator the fix is to either run
`TERM=xterm-256color && tmux`
or to set it to run the command as the login shell.
As for gnome dropdown terminal, it crashes when trying the above command. 
Either way the problem seems to be with the terminals them selves, not your config file. Thanks for the help! Closing.

**Comment by ddulic** [comment]
Reopening cause I found out that I actually have no color in my bash, I have colors down in tmux however, is this a part of your script? 
When I say colors I am referring to PS1, Directory colors, etc...

**Comment by gpakosz** [comment]
I would say this has nothing to do with this configuration.

**Comment by gpakosz** [comment]
Well now that I think about it, tmux manual says
>  default-shell path
>    Specify the default shell.  This is used as the login shell for new windows when the default-command option is set to empty, and must be the full path of the executable.  When started tmux tries to set a default value from the first suitable of the SHELL environment variable, the shell returned by getpwuid(3), or /bin/sh.  This option should be configured when tmux is used as a login shell.

And I'm making use of `$SHELL` when setting `default-command`. So you may want to check that `$SHELL` is also properly defined.

**Comment by gpakosz** [comment]
Or use your `~/.tmux.conf.local` copy to define `default-shell`

**Comment by gpakosz** [comment]
Please give the `default-command` branch a try and near the end of your `~/.tmux.conf.local` copy, add something like `set -g default-shell /bin/bash`

**Comment by ddulic** [comment]
Gave it a try. Didn't work. The output if `$SHELL` is` /bin/bash`, so that is correct, reloaded, killed all tmux instances, tried in different terminals, all a no go. I am running Ubuntu 16.04. with Gnome if that is of some importance.

**Comment by gpakosz** [comment]
Does it work with another configuration?

**Comment by ddulic** [comment]
Nope. Doesn't work at home with Fedora Gnome nor at work with Ubuntu Gnome. I will test it out today on KDE Neon and report back.

**Comment by gpakosz** [comment]
So to be clear, you're trying to use tmux as your login shell? This is not something I have tried but as far as Google tells me people are rather amending their `~/.bashrc` with something like
```
[ -z "$TMUX" ] && tmux
```

**Comment by ddulic** [comment]
Setting that in bash fixes the colors in the bottom in every terminal I have tried, however, the PS1 and Directory colors are still missing :/

**Comment by gpakosz** [comment]
And if you're using tmux without using my conf, are `PS1` etc correct?

**Comment by ddulic** [comment]
Without your conf, everything is correct.
![screenshot from 2016-11-30 11-08-53](https://cloud.githubusercontent.com/assets/10076428/20748209/9004f638-b6ed-11e6-853d-ec3ec6df6574.png)

**Comment by gpakosz** [comment]
Can you please take an equivalent screenshot with my conf?

**Comment by ddulic** [comment]
![screenshot from 2016-11-30 14-59-53](https://cloud.githubusercontent.com/assets/10076428/20755012/bb043554-b70d-11e6-9a2e-e40c9b6fc57c.png)

**Comment by gpakosz** [comment]
Is this `bash` ?

I just tried in Terminator, and I'm using `tmux -2` as a launch command and everything looks ok for me.

It looks like your `~/.bashrc` is not sourced properly. Can you make sure it's sourced by e.g. putting `echo "~/.bashrc sourced!"` at the end of `~/.bashrc` ?

**Comment by ddulic** [comment]
Yes, it is bash. It was owned by root for some reason, seems I accidentally added sudo to vim when making the .bash. It sourced out even when owned by root.

![one](https://cloud.githubusercontent.com/assets/10076428/20756364/b626d726-b712-11e6-943d-a50f31d81f5b.png)
![two](https://cloud.githubusercontent.com/assets/10076428/20756359/ae2eb3e0-b712-11e6-8dc6-97bc3dd99c20.png)
![three](https://cloud.githubusercontent.com/assets/10076428/20756355/ac626db8-b712-11e6-8dd5-73652a2bf990.png)

Ignore the 6 tmux instances, couldn't kill when taking the screenshot because I was doing something in another tab but I killed all after and its still the same :/

**Comment by gpakosz** [comment]
Can you screenshot your terminal settings as well, so that I can understand how you're launching `tmux` ?

**Comment by ddulic** [comment]
Just using the default gnome terminal, nothing changed, just the theme in general (Arc Theme) but that shouldn't effect anything, any particular setting you are interested in? I am using the bash configuration from [HERE](http://www.itworld.com/article/2833199/it-management/3-ways-to-pimp-your-bash-console.html) if that is of any help.

**Comment by gpakosz** [comment]
I'm interesting in how you launch `tmux`.

**Comment by ddulic** [comment]
At the moment I use your solution from top `[ -z "$TMUX" ] && tmux`
And I tried launching via the `tmux` command, and inserting it as a custom launch command in terminal settings. The result is the **same** no matter what method I use

**Comment by gpakosz** [comment]
Alright it's all about the way you customized your `~/.bashrc`. You need to make sure applies colors when `TERM` is `screen-256color`.

**Comment by ddulic** [comment]
Thanks for the help with this, but I am afraid I don't understand your answer. Do I need to set `TERM=screen-256color` after tmux is launched?

**Comment by gpakosz** [comment]
Generally speaking, for colors to work properly in `tmux`, `TERM` must be set to `xterm-256color` outside of `tmux`. And then in `.tmux.conf` people do `set -g default-terminal "screen-256color"`. This is already done by my conf so right now inside a `tmux` session `echo $TERM` should already report `screen-256color`.

Here's what's (likely, provided you copy pasted the conf from the link you mentioned) happening in your case:
  - your terminal emulator spawns `bash` and `TERM` is either `xterm` or `xterm-256color` depending on the configuration of the terminal emulator
  - `~/.bashrc` is sourced and applies colors
  - at the end of `~/.bashrc`, `[ -z "$TMUX" ] && tmux` spawns `tmux`
  - every time `tmux` creates a pane, it spawns `bash`
    - but this time `TERM` is set to `screen-256color`
    - and the configuration you copy pasted from doesn't take into account the case when `TERM` is set to `screen-256color`

All in all, I would use the following `~/.bashrc`

```
# outside of tmux, make sure TERM is xterm-256color
[ -z "$TMUX" ] && export TERM=xterm-256color

...

# customize colors
# make sure you handle both TERM=xterm-256color and TERM=screen-256color cases

...

# if not inside tmux already, replace current bash process by tmux
[ -z "$TMUX" ] && exec tmux
```

Hope that helps.

**Comment by ddulic** [comment]
Cheers. That did the trick! The whole problem by the way was that in `/etc/DIR_COLORS` I didn't have `TERM=screen-256color` set and
```
[ -z "$TMUX" ] && export TERM=xterm-256color
[ -z "$TMUX" ] && exec tmux
```
additionally fix the status bar colors inside tmux. Thanks once again for the plugin and for  the excessive help! :+1:
